### DISCUSSION TITLE
Improving the developer experience: adding the patch history

### DISCUSSION THREAD

**Discussion by uazo** [bromite/bromite#1239] [Ideas]
To date, after a year of development for bromite, I personally feel the lack of:
- be able to easily view the history of a patch (useful in reviews and regression studies)
- have the (more or less) certainty that the patches produced are applicable in the various versions, or, in any case, have the notification and the time necessary to make a patch of the patch :)

with regard to the last point a c.i. that also acts on canary could help, although the approach to the problem of how to apply the patches and keep their history is complex.

some hypotheses:

a) use of tools such as https://wiki.debian.org/UsingQuilt or https://stacked-git.github.io

I have tried stacked-git and it is very promising, but unfortunately it bases its functioning on the "automatic movement" of the
patch in a "stack", putting the patch you are working on on top, before working on it.
this is very inconvenient because:
- without the use of tools (like mtools) a move would touch many unnecessary files, lengthening the build times
- often, in development, it is easier to work on several points (therefore on several patches) at the same time and only afterwards do a final squash by assigning changes to the various patches

b) check if it is possible to create a repo organized in this way:
```
bromite:                                      --- bromite v90              --- bromite v91
                                             /                            /
                                            /                            /
branch patch A:            --- patch A --- /    ---->     ---- rev 1 ---/
                          /               /              /             /
branch patch B:          /---  patch B --/    ---->     /--- rev 1 ---/
                        /                              /
master chromium : v90 --                         v91 --
```

not having advanced knowledge of git I do not know if and how it is possible to set it up, nor if visualization tools allow me,
working on the bromite branch, to be able to see the history of the other branches (although it may not be a problem),
nor if the current structure of the patches (which also share the modification of the same files) would allow me to create a similar situation.

furthermore this could anyway
- do not solve the need for squash (or even complicate them), unless it is possible to create an automatic and reliable tool for generating patches,
- it would not simplify the application of the patches for the various versions of chromium.

c) evaluate the approach used by other downstream

to date I could only consider brave. vivaldi also exposes the sources but I haven't downloaded them yet, I don't know if other forks have the sources and build process open.

the approach that uses brave is complex and interesting: from what I understand they apply a series of patches to chromium that do not modify code but allow, using the C++ preprocessor, to be able to graft code from other source.

example:

https://github.com/brave/brave-core/blob/master/patches/base-threading-thread_restrictions.h.patch
```
--- a/base/threading/thread_restrictions.h
+++ b/base/threading/thread_restrictions.h
@@ -453,6 +453,7 @@ INLINE_IF_DCHECK_IS_OFF void DisallowBaseSyncPrimitives()
     EMPTY_BODY_IF_DCHECK_IS_OFF;
 
 class BASE_EXPORT ScopedAllowBaseSyncPrimitives {
+  BRAVE_SCOPED_ALLOW_BASE_SYNC_PRIMITIVES_H
  private:
   // This can only be instantiated by friends. Use
   // ScopedAllowBaseSyncPrimitivesForTesting in unit tests to avoid the friend
``` 

https://github.com/brave/brave-core/blob/master/chromium_src/base/threading/thread_restrictions.h
```
#ifndef BRAVE_CHROMIUM_SRC_BASE_THREADING_THREAD_RESTRICTIONS_H_
#define BRAVE_CHROMIUM_SRC_BASE_THREADING_THREAD_RESTRICTIONS_H_

class BraveBrowsingDataRemoverDelegate;

#define BRAVE_SCOPED_ALLOW_BASE_SYNC_PRIMITIVES_H \
  friend class ::BraveBrowsingDataRemoverDelegate;

#include "../../../../base/threading/thread_restrictions.h"
#undef BRAVE_SCOPED_ALLOW_BASE_SYNC_PRIMITIVES_H

#endif  // BRAVE_CHROMIUM_SRC_BASE_THREADING_THREAD_RESTRICTIONS_H_
```
unfortunately it is not always possible to do this, for example here they fall back into the old way:
https://github.com/brave/brave-core/blob/master/patches/components-safe_browsing-core-browser-safe_browsing_network_context.cc.patch
```
diff --git a/components/safe_browsing/core/browser/safe_browsing_network_context.cc b/components/safe_browsing/core/browser/safe_browsing_network_context.cc
index 1a3e3a87f17d8391f23edea4b45dc3a000bc4644..23617cd428c7e072c87cbfd9978e1c3562fd1aec 100644
--- a/components/safe_browsing/core/browser/safe_browsing_network_context.cc
+++ b/components/safe_browsing/core/browser/safe_browsing_network_context.cc
@@ -76,8 +76,8 @@ class SafeBrowsingNetworkContext::SharedURLLoaderFactory
       override {
     DCHECK(CurrentlyOnThread(ThreadID::UI));
     GetURLLoaderFactory()->CreateLoaderAndStart(
-        std::move(loader), request_id, options, request, std::move(client),
-        traffic_annotation);
+        std::move(loader), request_id, options, network::SystemRequestHandler::GetInstance()->OnBeforeSystemRequest(request),
+        std::move(client), traffic_annotation);
   }
 
   void Clone(mojo::PendingReceiver<network::mojom::URLLoaderFactory> receiver)
```
the advantage I imagine is a simplification of the rebase process for each version of chromium (perhaps also done by other people who do not necessarily have an understanding of the complexity of the base code, however unrelated to actual development and therefore probably pre-verified with weeks advance on canary which considering https://blog.chromium.org/2021/03/speeding-up-release-cycle.html is a nice plus) and, since the "real" code is on a different repo, keep its history.

for java, the approach is similar: it seems to me that they are exploiting the automatic modification of the bytecode through tools already present in the chromium toolchain.
they probably also have other mechanisms that have not yet been found or not made public.

to date I still do not have the ability to evaluate how much this approach is efficient. at first glance, without the ability to easily read the sum of the changes between patch and applied code, seems to be only disadvantageous, but I assume that they have some tools based perhaps on kythe to simplify reading.
anyway, considering they go on like this, I guess there are more advantages than disadvantages, or maybe it's the approach recommended by google itself for partners .. who knows.

I'd be curious to see how edge does it, but unfortunately the code is not in the public domain, although maybe something can be reconstructed backwards with what they publish, I don't know.

an obvious disadvantage I think is that the introduction of new employees requires a much more thorough preparation, certainly the team must be more prepared than necessary to modify bromite where the modifications are well separated and circumscribed.
i think this was one of the reasons why i was able to quickly modify the bromite base code.
in any case, the hypothesis of not increasing the necessary learning curve, is a point that should be considered.

in conclusion, the idea is to try to develop an evolution of the logic of brave, that try to keep the changes separated by functionality (as in bromite) but moving the patches that modify functions from those they modify upstream code (as in brave).
Unfortunately, the initial idea of having everything in new components I realized is not always feasible.

the question is very complex and I am still at the idea level, I still have nothing to show.
the direction I am checking is based on the possibility of modifying the chromium toolchain, for example by inserting a preprocessor for java (ref https://programming.guide/java/macros-in-java-source-files.html) which
can simulate the partial classes of C# and to investigate semantic patches (ref coccinelle and coccinelle4j).

Points of interest:
https://source.chromium.org/chromium/chromium/src/+/main:build/config/android/internal_rules.gni
https://source.chromium.org/chromium/chromium/src/+/main:build/android/gyp/
https://source.chromium.org/chromium/chromium/src/+/main:build/android/bytecode/

If you have any ideas i'm are all ears, I think the first step may be to do a bromite patch analyzer along their history,
to understand how they overlap, if they can be aggregated and how they have evolved over time and in the upstream code.

I keep you updated.

**Comment by csagan5** [comment]
> I assume that they have some tools based perhaps on kythe to simplify reading.

We could ask @wchen342, if he is at freedom to disclose this.

**Reply by wchen342** [reply]
Actually it is because of the workflow. Basically what happens is a problem comes up -> someone make a PR -> go through review -> close a issue /w the PR. It is kind of anti-pattern I guess, but that is how they work.
